src/node_worker.cc

```diff
@@ -56,6 +70,28 @@ Worker::Worker(Environment* env,
   Debug(this, "Creating new worker instance with thread id %llu",
         thread_id_.id);
 
+  active_worker_count++;
+  if (active_worker_count_warning &&
+      active_worker_count > per_process::cli_options->max_worker_thread_count) {
+    // Having too many active worker threads can degrade overall
+    // performance of the entire Node.js application by causing
+    // too much CPU contention. The default max-worker-threads is
+    // one less than 4 times the total number of CPUs available
+    // but may be set explicitly using the --max-worker-threads=n
+    // command line option.
+    // The warning below is emitted every time the worker count
+    // exceeds the limit. This is tracked per process rather than
+    // per environment/isolate so we can account also for all
+    // Workers created within Workers.
+    active_worker_count_warning = false;
+    USE(ProcessEmitWarning(env,
+                       "Too many active worker threads (%" PRId64 "). "
+                       "Performance may be degraded. "
+                       "(--max-worker-threads=%" PRId64 ")",
+                       active_worker_count,
+                       per_process::cli_options->max_worker_thread_count));
```

Comment:
This warning should always be emitted on the main thread, otherwise the thread on which it ends up being emitted is relatively random.

You can use `SetImmediateThreadsafe()` to recursively schedule callbacks on the parent thread until you hit the main thread, then emit the warning from there.

Reply:
ah, right, good point.

Reply:
Updated! Would appreciate a look over to make sure it looks correct :-)